### PR TITLE
8335397: Improve reliability of TestRecursiveMonitorChurn.java

### PR DESCRIPTION
**Pull request by xmas92**
TestRecursiveMonitorChurn.java currently uses NMT to try and correlate the native memory increase with unwanted inflation.

Change to instead query the JVM for exact number of inflations via the Whitebox API. This allow us to both be more exact and less dependent on interactions with NMT.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335397](https://bugs.openjdk.org/browse/JDK-8335397): Improve reliability of TestRecursiveMonitorChurn.java (**Enhancement** - P4)

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) 🔄 Re-review required (review applies to [81fd7c07](https://git.openjdk.org/jdk/pull/19965/files/81fd7c075f1d35e9bf0449ea0cb98231303b74c7))
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) 🔄 Re-review required (review applies to [44279523](https://git.openjdk.org/jdk/pull/19965/files/442795230dae18c26a808ab452872c35771eb5a5))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19965/head:pull/19965` \
`$ git checkout pull/19965`

Update a local copy of the PR: \
`$ git checkout pull/19965` \
`$ git pull https://git.openjdk.org/jdk.git pull/19965/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19965`

View PR using the GUI difftool: \
`$ git pr show -t 19965`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19965.diff">https://git.openjdk.org/jdk/pull/19965.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19965#issuecomment-2199667741)

### CODE REVIEW

**Review by rkennke — APPROVED** [review]
Looks good to me! Thank you!

File: test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java [hunk 1]
```diff
@@ -46,50 +48,32 @@ synchronized void doSomethingElse() {
...
-            // -           Object Monitors (reserved=208, committed=208)
-            //                             (malloc=208 #1) (at peak)
-            //
-            // Without recursive locking support, this would look more like this:
-            // -           Object Monitors (reserved=20800624, committed=20800624)
-            //                             (malloc=20800624 #100003) (at peak)
+    public static volatile Monitor monitor;
+    public static void main(String[] args) {
+        if (WB.getIntVMFlag("LockingMode") == LM_MONITOR) {
+            throw new SkippedException("LM_MONITOR always infaltes. Invalid test.");
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java:58]: typo: inflates

File: test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java [hunk 2]
```diff
@@ -46,50 +48,32 @@ synchronized void doSomethingElse() {
...
+        final long post_monitor_count = WB.getInUseMonitorCount();
+        System.out.println("Postcount = " + post_monitor_count);
 
-            Pattern pat = Pattern.compile("- *Object Monitors.*reserved=(\\d+), committed=(\\d+).*");
-            for (String line : output.asLines()) {
-                Matcher m = pat.matcher(line);
-                if (m.matches()) {
-                    long reserved = Long.parseLong(m.group(1));
-                    long committed = Long.parseLong(m.group(2));
-                    System.out.println(">>>>> " + line + ": " + reserved + " - " + committed);
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java:85]: Oh so it just measures how much memory we use for ObjectMonitors?  yes, this doesn't seem very reliable.

**Review by coleenp — APPROVED** [review]
This is an improvement.

File: test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java (hunk 1 above)
Comment (xmas92) [reply test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java:58]:
```suggestion
            throw new SkippedException("LM_MONITOR always inflates. Invalid test.");
```

**Review by coleenp — APPROVED** [review]

File: src/hotspot/share/prims/whitebox.cpp [hunk 3]
```diff
@@ -1850,6 +1854,10 @@ WB_ENTRY(jboolean, WB_IsMonitorInflated(JNIEnv* env, jobject wb, jobject obj))
   return (jboolean) obj_oop->mark().has_monitor();
 WB_END
 
+WB_ENTRY(jlong, WB_getInUseMonitorCount(JNIEnv* env, jobject wb))
+  return (jlong) WhiteBox::get_in_use_monitor_count();
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/whitebox.cpp:1858]: Why the indirection?

File: test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java [hunk 4]
```diff
@@ -46,50 +48,32 @@ synchronized void doSomethingElse() {
...
-                    if (committed > 1000) {
-                        throw new RuntimeException("Allocated too many monitors");
-                    }
-                    return;
-                }
+        if (pre_monitor_count != post_monitor_count) {
+            final long monitor_count_change = post_monitor_count - pre_monitor_count;
+            System.out.println("Unexpected change in mointor count: " + monitor_count_change);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java:72]: typo: mointor

**Review by dholmes-ora — APPROVED** [review]
This seems much more reliable if we are testing the absence of excessive inflation.

One query below and one typo, but approved.

Thanks

File: src/hotspot/share/prims/whitebox.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/whitebox.cpp:1858]: Ah now I see. We need the member function to make it a friend.

File: test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java (hunk 4 above)
Comment (xmas92) [reply test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java:72]:
```suggestion
            System.out.println("Unexpected change in monitor count: " + monitor_count_change);
```

**Review by coleenp — APPROVED** [review]

**Comment by xmas92** [comment]
Thanks for the reviews.
/integrate

**Pull request closed without merging**